### PR TITLE
Add HDR to Metal + Perceptual HDR

### PR DESCRIPTION
**Pull request by Sam-Belliveau**
This PR:

1. Add HDR to Metal
3. Add PerceptualHDR

PerceptualHDR Menu:
image

### CODE REVIEW

**Comment by MayImilae** [comment]
> Likely not ready to be merged unless it shows some huge advantage as it just removes a feature atm.

You should probably mark this PR as draft then.

File: Data/Sys/Shaders/default_pre_post_process.glsl [hunk 1]
```diff
@@ -195,214 +255,169 @@ float4 SharpBilinearSample(float3 uvw, float gamma)
...
+            color.rgb = color.rgb * from_NTSCM;
+        else if (game_color_space == 1)
+            color.rgb = color.rgb * from_NTSCJ;
+        else if (game_color_space == 2)
+            color.rgb = color.rgb * from_PAL;
+    }
+
+    if (OptionEnabled(hdr_output)) {
+        float hdr_paper_white = hdr_paper_white_nits / hdr_sdr_white_nits;
+        color = OklabToLinearRGB(hdr_paper_white * LinearRGBToOklab(color));
```
Comment (Filoppi) [inline Data/Sys/Shaders/default_pre_post_process.glsl:415]: This is wrong, to scale up the brightness of Oklab colors, you should just scale its 3rd component (intensity, which is comparable to brightness/luminance).
Though even after that, it's a very unconventional and unnecessary way of scaling up the brightness from SDR to HDR.
Usually brightness is scaled linearly with linear space colors, and that's what the code was doing before. It's akin to changing the brightness of your monitor/TV in its settings, or to changing the Windows SDR to HDR brightness slider.
Generally speaking, there's no need to scale colors with Oklab to maintain the original perceived colors (the color our eyes would have perceived at the original SDR brightness, which anyway, wouldn't necessarily be 80 or 100 nits, but it would depend on the TVs the game was designed on).
This is likely to not make much difference because color perception doesn't change much between ~100 and (e.g.) 300 nits, but it's just unnecessary (IMO) and "slower". Unless you can come up with an explanation for doing this, let's avoid it. Even then, it should be a separate (HDR) post process, like my AutoHDR post process shader is.

Comment (Filoppi) [inline Data/Sys/Shaders/default_pre_post_process.glsl:415]: This is wrong: to scale up the brightness of Oklab colors, you should just scale its 3rd component (intensity, which is comparable to brightness/luminance).
Though even after that, it's a very unconventional and unnecessary way of scaling up the brightness from SDR to HDR, and the output brightness wouldn't match the target nits the user had selected (it wouldn't match at all, as Oklab intensity is not in linear space, so multiplying it yields very different results).
Usually brightness is scaled "linearly", with linear space colors, and that's what the code was doing before. It's akin to changing the brightness of your monitor/TV in its settings, or to changing the Windows SDR to HDR brightness slider.
Generally speaking, there's no need to scale colors with Oklab to maintain the original perceived colors (the color our eyes would have perceived at the original SDR brightness, which anyway, wouldn't necessarily be 80 or 100 nits, but it would depend on the TVs the game was designed on).
This is likely to not make much difference because color perception doesn't change much between ~100 and (e.g.) 300 nits, but it's just unnecessary (IMO) and "slower". Unless you can come up with an explanation for doing this, let's avoid it. Even then, it should be a separate (HDR) post process, like my AutoHDR post process shader is.

File: Data/Sys/Shaders/default_pre_post_process.glsl [hunk 2]
```diff
@@ -4,92 +4,152 @@
...
+
+        return c;
+    }
+
+    else {
+        return pow(c, float4(1.0 / sdr_display_custom_gamma));
+    }
+}
+
+float4 sRGBToLinear(float4 c)
```
Comment (Filoppi) [inline Data/Sys/Shaders/default_pre_post_process.glsl:47]: `LinearTosRGB()` `sRGBToLinear()` should be renamed to "GammaToLinear()" or "PerceptionSpaceToLinearSpace()", as sRGB is never used here (the sRGB gamma branch is disabled), and was never used by GameCube/Wii games AFAIK.

File: Data/Sys/Shaders/default_pre_post_process.glsl [hunk 3]
```diff
@@ -4,92 +4,152 @@
...
+
+float4 OklabToLinearRGB(float4 c)
+{
+    return max(LMStoRGB * pow(OklabtoLMS * c, float4(3.0)), 0.0);
+}
+
+/***** Quick Sampling *****/
 
 // Non filtered gamma corrected sample (nearest neighbor)
-float4 QuickSample(float3 uvw, float gamma)
```
Comment (Filoppi) [inline Data/Sys/Shaders/default_pre_post_process.glsl:44]: I'm not a big fan of removing the gamma parameter from samples and instead reading it from the global parameter.
Performance wise, nothing changes as shader functions are all inlined when compiled, but the functions are less generic and re-usable now, which I don't like. Unnecessary change IMO.

File: Data/Sys/Shaders/default_pre_post_process.glsl [hunk 4]
```diff
@@ -195,214 +255,169 @@ float4 SharpBilinearSample(float3 uvw, float gamma)
...
-	}
-
-	return color;
+    float3 uvw = v_tex0;
+
+    float4 color = float4(0.0, 0.0, 0.0, 1.0);
+
+    // Bilinear
+    if (resampling_method <= 1)
+        color = BilinearSample(uvw);
```
Comment (Filoppi) [inline Data/Sys/Shaders/default_pre_post_process.glsl:362]: I find not using parenthesis in branches that have spacing between each if and else unconventional

File: Data/Sys/Shaders/default_pre_post_process.glsl [hunk 5]
```diff
@@ -195,214 +255,169 @@ float4 SharpBilinearSample(float3 uvw, float gamma)
...
-	// Restore the original gamma without changes
-	else
-	{
-			color.rgb = pow(color.rgb, float3(1.0 / game_gamma));
-	}
-
-	SetOutput(color);
+    float4 color = float4(0.0, 0.0, 0.0, 1.0);
+
+    if (GetResolution() == GetWindowResolution() || resampling_method == 0)
```
Comment (Filoppi) [inline Data/Sys/Shaders/default_pre_post_process.glsl:399]: This new code removes all the comments that explained the logic of what was being done.
This branch was checking for `needs_resampling`:
```
bool raw_resampling = resampling_method <= 0;
bool needs_rescaling = GetResolution() != GetWindowResolution();

bool needs_resampling = needs_rescaling && (OptionEnabled(hdr_output) || OptionEnabled(correct_gamma) || !raw_resampling);
```
while now it does some inline checks that are not clearly why they are being made.

File: Data/Sys/Shaders/default_pre_post_process.glsl [hunk 6]
```diff
@@ -195,214 +255,169 @@ float4 SharpBilinearSample(float3 uvw, float gamma)
...
+        color = sRGBToLinear(texture(samp1, v_tex0));
+    else
+        color = GetResampledColor();
+
+    if (OptionEnabled(correct_color_space)) {
```
Comment (Filoppi) [inline Data/Sys/Shaders/default_pre_post_process.glsl:404]: Not a big deal, but dolphin codebase  doesn't use `if () {`, `{` should be on a new line.

**Review by Filoppi — CHANGES_REQUESTED** [review]
Overall I find this PR to have more negatives than positives at the moment.
I don't like most of the code restructuring, I find it unnecessary (because it's working, and syntax doesn't matter as much) and detrimental (because it adds more problems and removes comments and code that made the shader easier to read).

I'm not against spacing changes, so if you want to make a PR for that, feel free to do so.

Scaling brightness by oklab in HDR isn't really wanted here.
The original idea of blending texture samples with oklab could have been elaborated more into something better though, as long as we can get the most physically accurate (which is a bit different than perceptually accurate) color mixing.

**Comment by Sam-Belliveau** [comment]
This is hard to show because i cant take HDR screenshots and my phone adjusts brightness, but I need you to know that the results from this branch look significantly nicer than the existing algorithm.

HDR (500 paper white nits, does not look blown out on my display):
image

Default:
image

**Comment by Sam-Belliveau** [comment]
HDR:

image

SDR:
image

**Comment by Sam-Belliveau** [comment]
This one is hard to explain when they're both blown out, but the Oklab HDR is considerably more vibrant and colorful, along with having deeper blacks. (this is because the increase in brightness is slightly accounted for) 

It is definitely worth a shot and something new and exciting!

## Oklab HDR

image

## Standard HDR

image

File: Data/Sys/Shaders/default_pre_post_process.glsl (hunk 1 above)
Comment (Sam-Belliveau) [reply Data/Sys/Shaders/default_pre_post_process.glsl]: Scaling the 3rd component looks weird and desaturated. Scaling it like this increases perceived exposure.

File: Data/Sys/Shaders/default_pre_post_process.glsl [hunk 7]
```diff
@@ -195,214 +255,175 @@ float4 SharpBilinearSample(float3 uvw, float gamma)
...
+
+    if (OptionEnabled(hdr_output)) {
+        float hdr_paper_white = hdr_paper_white_nits / hdr_sdr_white_nits;
+        hdr_paper_white = pow(hdr_paper_white, 1.0 / 3.0); // Oklab is in a cube root space
+
+        // in order to use the full range of the display at full HDR,
+        // linear brighness needs to go down when the exposure goes up.
+        // This is because scRGB is a -0.5 to 7.5 range.
+        float4 brightness_adjustment = float4((hdr_paper_white - 1) / 6.0, 0.0, 0.0, 0.0);
+        color = OklabToLinearRGB(hdr_paper_white * LinearRGBToOklab(color) - brightness_adjustment);
```
Comment (Filoppi) [inline Data/Sys/Shaders/default_pre_post_process.glsl:421]: This is a very unconventional way of upgrading SDR to HDR and it's subjective. This needs to be a post process effect, akin to the AutoHDR post process shader that Dolphin already has.
This this will also create negative oklab values, due to the subtraction, which could result in very random/invalid colors.

This currently shifts both hue, chroma/saturation and intensity/brightness, thus making it a 100% non conservative SDR to HDR upgrade path. I suggest using Oklch to boost up the chroma and intensity only, without touching the hue, but even then, this is a bit random of a thing to do. I feel like you are just amazed by the boosted perceived saturation and brightness as of now, but that's not really what Dolphin goes for, that should be your own post process that you can use. The artistic intent of games is ruined with this change.

File: Source/Core/VideoBackends/Metal/MTLMain.mm [hunk 8]
```diff
@@ -119,8 +119,13 @@ static bool WindowSystemTypeSupportsMetal(WindowSystemType type)
 
     MRCOwned<CAMetalLayer*> layer = MRCRetain(static_cast<CAMetalLayer*>(wsi.render_surface));
     [layer setDevice:adapter];
-    if (Util::ToAbstract([layer pixelFormat]) == AbstractTextureFormat::Undefined)
-      [layer setPixelFormat:MTLPixelFormatBGRA8Unorm];
+    [layer setWantsExtendedDynamicRangeContent:YES];
```
[the remark this answers is not in the document]
Comment (Filoppi) [reply Source/Core/VideoBackends/Metal/MTLMain.mm:122]: We should check if the scRGB HDR white point on mac os is 80 or 100 nits. On Windows it's 80 nits.
Like, a texture value of 1 1 1 (white), what brightness will it output on your display on mac os?

**Review by Filoppi — CHANGES_REQUESTED** [review]

File: Data/Sys/Shaders/default_pre_post_process.glsl (hunk 7 above)
Comment (Sam-Belliveau) [reply Data/Sys/Shaders/default_pre_post_process.glsl:421]: Multiplying brightness and chroma in Oklch by an amount is mathematically identical to multiplying all 3 axis in Oklab by that amount. This does not introduce any hue shift and IS a conservative upgrade path. Negative Oklab colors are actually ok, they don't look bad, they just look darker.

I encourage you to actually try this out, the artistic intent of the game is not lost, and it looks like HDR is supposed to. I've shown many people in person and they all agree this is an improvement.

Comment (Filoppi) [reply Data/Sys/Shaders/default_pre_post_process.glsl:421]: you are right, there is no hue shift, just chroma and brightness. 
I tried it. What I said still applies, this a subjective SDR to HDR upgrade, and thus should be in a post process shader file.

**Comment by Filoppi** [comment]
Please just implement this as a post process, just like `AutoHDR.glsl` that's where it belongs with the way things are design.
You are applying a subjective SDR to HDR upgrade, and that should not be part of the default feature set of "color correction", as it's something "random", based on taste.

**Comment by Sam-Belliveau** [comment]
I really disagree with the idea that this is something random based on taste. I showed before my exact derivation, reasoning, and verification that it is not distorting the underlying picture. 

I explained exactly the problem it's solving. (dark colors becoming over exposed) I explain the requirements for an algorithm to resolve it, and I picked the single function that exists that solves this problem. There are no tunable parameters for the perceptual mode, because only one solution exists. 

It fits the requirements for HDR Paper Nits. White is the exact same as in the Linear mode. Great lengths were made on my end to make sure this was more than just a random feature. 

It was part of a larger effort undertaken to clean up this default post process shader. This is more than just a post process function, this is an entirely valid, mathematically backed way to scale SDR to HDR given certain constraints. 

This is the form that I believe this feature should be in, because it's simple and offers substantive improvements. It would take extra work for me to go through my refactor and undo all and make it only support a single form of HDR Scaling. 

I'd be willing to consider a post process effect in the future but this is the state of this project right now.

File: Data/Sys/Shaders/default_pre_post_process.glsl (hunk 3 above)
Comment (iwubcode) [reply Data/Sys/Shaders/default_pre_post_process.glsl:44]: I would agree, I'd like to keep these reusable if possible

**Comment by iwubcode** [comment]
I am not as passionate as either of you about color output, HDR, or gamma.  Regardless, I will give some general thoughts /opinions.

Maybe @Filoppi can remind me, why is **HDR Paper White Nits** even an option for the color spaces?  To me it seems like an odd option to have in a color correction window about the original GC/Wii.  From what I recall, you said that the HDR range was needed to show those color spaces.  Can we not determine the appropriate white balance for those original spaces?

Even calling this window "color correction" gives off the wrong expectations.  This window isn't about color correction or HDR, that is a post process function.  Therefore I would lean toward not including anything that isn't related to our original goal (console accuracy) in this window.  (Additionally, I'd like to avoid making this window too large)

Also, while it is great we cleaned the shader up, I don't think we should get too attached to our current shader setup for post processing.  I can't make any promises but I'd like to rework shader management in the future (may or may not be a part of post processing enhancement).

Having done a number of ports of Reshade shaders and also worked on a number of custom shaders with my new material system, it's apparent that our current approach doesn't scale very well.  Our global functions are difficult to find without reading through a lot of source and users get everything we put in global state when they implement their own shader.  We really need some common code approach that I'll call a "library".

Sorry this was a long amount of text.  I'll give it some more thought but my initial impression is that OkLab would be good to include in a Dolphin provided "library" but the perceptual color conversion makes more sense in a post processing shader that the user can opt-in to use.

**Review by iwubcode — APPROVED** [review]
Shader part LGTM.

File: Source/Core/VideoBackends/Metal/MTLMain.mm [hunk 9]
```diff
@@ -119,8 +119,13 @@ static bool WindowSystemTypeSupportsMetal(WindowSystemType type)
...
+    [layer setPixelFormat:MTLPixelFormatRGBA16Float];
+
+    const CFStringRef name = kCGColorSpaceExtendedLinearSRGB;
+    CGColorSpaceRef colorspace = CGColorSpaceCreateWithName(name);
+    [layer setColorspace:colorspace];
+    CGColorSpaceRelease(colorspace);
```
Comment (TellowKrinkle) [inline Source/Core/VideoBackends/Metal/MTLMain.mm:128]: SDR mode is noticeably different with this change (left is before this change, right is after):
image

Also, using RGBA16 disables full screen optimizations on my Intel mac (`MTL_HUD_ENABLED=1`, watch for whether it says composited or direct when fullscreen).  You probably only want to do this if the user requested HDR.

File: Data/Sys/Shaders/PerceptualHDR.glsl [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+const mat4 LMStoOklab = mat4(
+    0.2104542553, 1.9779984951, 0.0259040371, 0.0000000000,
+    0.7936177850, -2.4285922050, 0.7827717662, 0.0000000000,
+    -0.0040720468, 0.4505937099, -0.8086757660, 0.0000000000,
+    0.0000000000, 0.0000000000, 0.0000000000, 1.0000000000);
+
+float4 LinearRGBToOklab(float4 c)
+{
+    return LMStoOklab * pow(RGBtoLMS * c, float4(1.0 / 3.0));
```
Comment (Filoppi) [inline Data/Sys/Shaders/PerceptualHDR.glsl:31]: Any reason you are also converting the alpha channel?

**Review by Filoppi — APPROVED** [review]
Other than the few comments left, code looks good now. Thanks for making this a post process shader.

**Review by TellowKrinkle — APPROVED** [review]
Metal part LGTM
(Though I'd prefer if the Metal changes were in a separate commit from the PerceptualHDR shader)

File: Data/Sys/Shaders/PerceptualHDR.glsl (hunk 10 above)
Comment (Sam-Belliveau) [reply Data/Sys/Shaders/PerceptualHDR.glsl:39]: It is unaffected throughout this process. It is reversed at the end.
